File: src/ai-module/app/api/api_v1/endpoints/entities.py
"""
Entity API endpoints for the AI Development Module.

This module provides REST API endpoints for entity management including
CRUD operations, search, and extraction from documents.
"""

import uuid
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from ....core.database import get_db
from ....models import Entity
from ....schemas.entity import (
    EntityCreate,
    EntityResponse,
    EntityUpdate,
    EntityListResponse,
    EntityExtractionRequest,
    EntityExtractionResponse,
)

# Create router
router = APIRouter()


@router.post("/", response_model=EntityResponse)
async def create_entity(
    entity: EntityCreate,
    db: AsyncSession = Depends(get_db)
) -> EntityResponse:
    """Create a new entity."""
    # Check if entity with same name and type already exists
    existing = await db.execute(
        "SELECT * FROM entities WHERE name = :name AND type = :type AND is_active = true",
        {"name": entity.name, "type": entity.type}
    )
    if existing.first():
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Entity with name '{entity.name}' and type '{entity.type}' already exists"
        )

    # Create new entity
    db_entity = Entity(
        name=entity.name,
        type=entity.type,
        confidence=entity.confidence,
        source=entity.source,
        source_type=entity.source_type,
        source_document_id=entity.metadata.get("source_document_id") if entity.metadata else None,
        extraction_method="manual",  # Manual creation via API
        positions=entity.positions,
        metadata=entity.metadata or {},
        vector_embedding=None,  # Will be generated asynchronously
    )

    db.add(db_entity)
    await db.commit()
    await db.refresh(db_entity)

    return EntityResponse.from_orm(db_entity)


@router.get("/", response_model=EntityListResponse)
async def list_entities(
    entity_type: Optional[str] = Query(None, description="Filter by entity type"),
    confidence_min: Optional[float] = Query(None, ge=0.0, le=1.0, description="Minimum confidence score"),
    source: Optional[str] = Query(None, description="Filter by source"),
    limit: int = Query(50, ge=1, le=100, description="Results per page"),
    offset: int = Query(0, ge=0, description="Pagination offset"),
    db: AsyncSession = Depends(get_db)
) -> EntityListResponse:
    """List entities with filtering and pagination."""

    # Build query
    query = "SELECT * FROM entities WHERE is_active = true"
    params = {}

    if entity_type:
        query += " AND type = :entity_type"
        params["entity_type"] = entity_type

    if confidence_min is not None:
        query += " AND confidence >= :confidence_min"
        params["confidence_min"] = confidence_min

    if source:
        query += " AND source = :source"
        params["source"] = source

    query += " ORDER BY created_at DESC LIMIT :limit OFFSET :offset"
    params.update({"limit": limit, "offset": offset})

    # Get total count
    count_query = query.replace("SELECT *", "SELECT COUNT(*)", 1).replace(
        "ORDER BY created_at DESC LIMIT :limit OFFSET :offset", ""
    )
    total_result = await db.execute(count_query, params)
    total_count = total_result.scalar()

    # Get entities
    result = await db.execute(query, params)
    entities = result.fetchall()

    # Calculate pagination info
    has_more = (offset + limit) < total_count

    return EntityListResponse(
        entities=[EntityResponse.from_orm(entity) for entity in entities],
        total_count=total_count,
        page=(offset // limit) + 1,
        page_size=limit,
        has_more=has_more
    )


@router.get("/{entity_id}", response_model=EntityResponse)
async def get_entity(
    entity_id: str,
    db: AsyncSession = Depends(get_db)
) -> EntityResponse:
    """Get entity by ID."""
    result = await db.execute(
        "SELECT * FROM entities WHERE id = :entity_id AND is_active = true",
        {"entity_id": entity_id}
    )

    entity = result.first()
    if not entity:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entity with ID {entity_id} not found"
        )

    return EntityResponse.from_orm(entity)


@router.put("/{entity_id}", response_model=EntityResponse)
async def update_entity(
    entity_id: str,
    entity_update: EntityUpdate,
    db: AsyncSession = Depends(get_db)
) -> EntityResponse:
    """Update an existing entity."""
    # Get existing entity
    result = await db.execute(
        "SELECT * FROM entities WHERE id = :entity_id AND is_active = true",
        {"entity_id": entity_id}
    )

    entity = result.first()
    if not entity:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entity with ID {entity_id} not found"
        )

    # Update fields
    update_data = entity_update.dict(exclude_unset=True)
    if update_data:
        for field, value in update_data.items():
            setattr(entity, field, value)

        await db.commit()
        await db.refresh(entity)

    return EntityResponse.from_orm(entity)


@router.delete("/{entity_id}")
async def delete_entity(
    entity_id: str,
    db: AsyncSession = Depends(get_db)
) -> dict:
    """Soft delete an entity (mark as inactive)."""
    result = await db.execute(
        "SELECT * FROM entities WHERE id = :entity_id AND is_active = true",
        {"entity_id": entity_id}
    )

    entity = result.first()
    if not entity:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entity with ID {entity_id} not found"
        )

    # Soft delete
    entity.is_active = False
    await db.commit()

    return {"message": f"Entity {entity_id} deleted successfully"}


@router.post("/extract", response_model=EntityExtractionResponse)
async def extract_entities(
    request: EntityExtractionRequest,
    db: AsyncSession = Depends(get_db)
) -> EntityExtractionResponse:
    """Extract entities from document content."""

    # TODO: Implement actual entity extraction logic
    # For now, return a mock response

    job_id = str(uuid.uuid4())

    # Create processing job
    from ....models import DocumentProcessingJob

    job = DocumentProcessingJob(
        document_id=request.document_id,
        status="processing",
        priority=request.priority,
        extraction_config=request.extraction_config.dict(),
    )

    db.add(job)
    await db.commit()

    # Mock entity extraction results
    mock_entities = [
        {
            "id": str(uuid.uuid4()),
            "name": "Example Corp",
            "type": "organization",
            "confidence": 0.95,
            "source": request.document_id or "manual",
            "source_type": request.document_type,
            "extraction_method": "llm",
            "metadata": {},
            "positions": [[0, 12]],
            "created_at": "2025-01-23T12:00:00Z",
            "updated_at": "2025-01-23T12:00:00Z",
            "is_active": True
        }
    ]

    return EntityExtractionResponse(
        job_id=job_id,
        status="completed",
        entities=mock_entities,
        relationships=[],
        processing_time_seconds=2.5
    )

File: frontend/src/mocks/handlers/entities.ts
import { http, HttpResponse } from 'msw'
import { mockEntities, searchEntities, getEntitiesByType } from '../data/entities'
import { faker } from '@faker-js/faker'
import type { Entity } from '@/types'

export const entityHandlers = [
  // Get entities with pagination and filters
  http.get('/api/v1/entities', async ({ request }) => {
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, faker.number.int({ min: 100, max: 300 })))
    
    const url = new URL(request.url)
    const limit = parseInt(url.searchParams.get('limit') || '20')
    const offset = parseInt(url.searchParams.get('offset') || '0')
    const entityType = url.searchParams.get('entity_type')
    const confidenceMin = parseFloat(url.searchParams.get('confidence_min') || '0')
    const source = url.searchParams.get('source')
    const searchQuery = url.searchParams.get('q')
    
    // Start with all active entities
    let filtered = mockEntities.filter(e => e.is_active)
    
    // Apply filters
    if (entityType) {
      filtered = filtered.filter(e => e.type === entityType)
    }
    
    if (confidenceMin > 0) {
      filtered = filtered.filter(e => e.confidence >= confidenceMin)
    }
    
    if (source) {
      filtered = filtered.filter(e => e.source_type === source)
    }
    
    if (searchQuery) {
      const lowerQuery = searchQuery.toLowerCase()
      filtered = filtered.filter(e => 
        e.name.toLowerCase().includes(lowerQuery) ||
        (e.metadata && JSON.stringify(e.metadata).toLowerCase().includes(lowerQuery))
      )
    }
    
    // Paginate
    const total = filtered.length
    const paginated = filtered.slice(offset, offset + limit)
    
    return HttpResponse.json({
      data: paginated,
      pagination: {
        page: Math.floor(offset / limit) + 1,
        page_size: limit,
        total_count: total,
        has_more: offset + limit < total
      },
      execution_time_ms: faker.number.int({ min: 50, max: 200 })
    })
  }),
  
  // Get single entity by ID
  http.get('/api/v1/entities/:id', async ({ params }) => {
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, 150))
    
    const entity = mockEntities.find(e => e.id === params.id)
    
    if (!entity) {
      return HttpResponse.json(
        {
          error: 'Not found',
          message: 'Entity not found'
        },
        { status: 404 }
      )
    }
    
    return HttpResponse.json(entity)
  }),
  
  // Create new entity
  http.post('/api/v1/entities', async ({ request }) => {
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, 400))
    
    const body = await request.json() as Partial<Entity>
    
    // Validate required fields
    if (!body.name || !body.type) {
      return HttpResponse.json(
        {
          error: 'Validation error',
          message: 'Name and type are required'
        },
        { status: 400 }
      )
    }
    
    // Create new entity
    const newEntity: Entity = {
      id: faker.string.uuid(),
      name: body.name,
      type: body.type,
      confidence: body.confidence || 0.8,
      source: body.source || 'manual_entry',
      source_type: body.source_type || 'manual',
      extraction_method: 'manual',
      positions: body.positions,
      metadata: body.metadata || {},
      vector_embedding: body.vector_embedding,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
      is_active: true
    }
    
    // Add to mock entities
    mockEntities.push(newEntity)
    
    return HttpResponse.json(newEntity, { status: 201 })
  }),
  
  // Update entity
  http.put('/api/v1/entities/:id', async ({ params, request }) => {
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, 350))
    
    const body = await request.json() as Partial<Entity>
    const index = mockEntities.findIndex(e => e.id === params.id)
    
    if (index === -1) {
      return HttpResponse.json(
        {
          error: 'Not found',
          message: 'Entity not found'
        },
        { status: 404 }
      )
    }
    
    // Update entity
    mockEntities[index] = {
      ...mockEntities[index],
      ...body,
      id: mockEntities[index].id, // Don't allow ID changes
      updated_at: new Date().toISOString()
    }
    
    return HttpResponse.json(mockEntities[index])
  }),
  
  // Delete entity (soft delete)
  http.delete('/api/v1/entities/:id', async ({ params }) => {
    // Simulate network delay
    await new Promise(resolve => setTimeout(resolve, 250))
    
    const index = mockEntities.findIndex(e => e.id === params.id)
    
    if (index === -1) {
      return HttpResponse.json(
        {
          error: 'Not found',
          message: 'Entity not found'
        },
        { status: 404 }
      )
    }
    
    // Soft delete
    mockEntities[index].is_active = false
    mockEntities[index].updated_at = new Date().toISOString()
    
    return HttpResponse.json({ success: true, message: 'Entity deleted' })
  }),
  
  // Extract entities from text (mock AI extraction)
  http.post('/api/v1/entities/extract', async ({ request }) => {
    // Simulate longer processing for AI extraction
    await new Promise(resolve => setTimeout(resolve, faker.number.int({ min: 800, max: 1500 })))
    
    const body = await request.json() as { content: string; document_type?: string }
    
    // Mock extraction: return random entities
    const extractedCount = faker.number.int({ min: 3, max: 10 })
    const extractedEntities = faker.helpers.arrayElements(mockEntities, extractedCount)
    
    return HttpResponse.json({
      job_id: faker.string.uuid(),
      entities: extractedEntities,
      relationships: [], // Would include relationships in real implementation
      processing_time_ms: faker.number.int({ min: 800, max: 1500 })
    })
  })
]














